Implementation Plan: Managing A Sensitive Terms List (#911)
WordPress/openverse: rfcs/detecting_sensitive_textual_content/20230309-implementation_plan_sensitive_terms_list.md

# Implementation Plan: Managing A Sensitive Terms List

> **Content Warning**
>
> This document **does not** include sensitive content or direct links to
> sensitive content. It _does_ discuss different sub-categorizations of
> sensitive content at times.
>
> This document should not be considered sensitive but please view at your own
> discretion, regardless.

## Reviewers

<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->

- [ ] @sarayourfriend
- [ ] @dhruvkb

The Openverse project
[Filter and blur sensitive results by term matching #377](https://github.com/orgs/WordPress/projects/70?query=is%3Aopen+sort%3Aupdated-desc&pane=issue&itemId=19530115)
requires that we compare the textual contents of Openverse media against a

> denylist of sensitive terms. We use this list to identify media that should be

Reviewer: Just a nit that "denylist" might imply that we _are_ rejecting these documents (despite the qualification in the following sentence).

Author: Good point. Switched to "list" in 950bc718.

marked as "sensitive," which means they should not be displayed to users by
default, but only with their explicit consent. Importantly, this list is not
used to censor or deny access to specific works, but to enable opt-in access.

The list must be accessible to contributors who need to modify it, but not so
public that it could be seen by users who would rather not be exposed to it. We
recognize that the list itself is sensitive content, and it needs to be treated
with care.

> To create and manage this list I reccommend the following guidelines.

Reviewer: ```suggestion
To create and manage this list I recommend the following guidelines.
```

Author: Fixed in 950bc718.

    Spanish, Brazilian Portugese, Turkish, and Russian. English is the only
    language used by all maintainers.

3.  On what basis will Openverse maintainers use their "discretion" to create
    this list?

    Until our community has developed its own standards around sensitivity, we
    can use a few existing documents in the WordPress and user-submitted content
    ecosystem to guide us:

    - WordPress's page on [etiquette](https://wordpress.org/about/etiquette/) in
      the project.
    - WordPress's
      [code of conduct](https://make.wordpress.org/community/handbook/wordcamp-organizer/planning-details/code-of-conduct/)
      for in-person events.
    - YouTube's
      [community guidelines](https://www.youtube.com/howyoutubeworks/policies/community-guidelines/)
      offers a decent list of categories for unacceptable types of content. For
      example, their
      [thumbnail policy](https://support.google.com/youtube/answer/9229980?hl=en&ref_topic=9282679#zippy=%2Cage-restricted-thumbnails-and-thumbnail-removal)
      includes some information on types of sexual and violent content that
      either can't be used or can be used behind a flag comparable to ours.

>       - Additionally, see the [#this-will-not-be-perfect](#this-wont-be-perfect)
>         section for additional

Reviewer: Editorial.

```suggestion
    - Additionally, see the [#this-will-not-be-perfect](#this-will-not-be-perfect)
      section for additional
```

Fixing the indentation here (maybe?) but also noting that the phrase appears incomplete.

Perhaps just deleting the "section for additional" would be fine. In context it makes sense with the name of the section referenced:

```suggestion
  - Additionally, see the [#this-will-not-be-perfect](#this-will-not-be-perfect)
```

I've also updated the heading reference in the link location as it was using `wont` instead of `will-not` and didn't match the actual heading.

Author: Fixed in 950bc718, thanks.

## Implementation Plan

1. Ping a WordPress GitHub account maintainer to request that they create the
   `WordPress/openverse-sensitive-terms` repository.
2. Initialize the repository with a `README.md` and a blank
   `sensitive-terms.txt` file.
3. Create a PR which initializes the sensitive terms list by copying all terms
   in the "List of Dirty, Naughty, Obscene, and Otherwise Bad Words" repository.
   _I am not including the link here to prevent accidental viewing of the
   repository, but feel free to find it within the GitHub organization
   "LDNOOBW"._
4. Identify willing reviewers from the `@WordPress/openverse-maintainers` group
   who are comfortable with looking at the list and reviewing the PR, looking
   for terms which should be removed or missing terms which should be added.
5. Merge the PR, and now the list is available for use in Openverse projects.
6. Publish a "Make" post about the creation of the list and its purpose.

> 

Reviewer: Can this implementation plan include how the sensitive terms list will be used in the Terraform repository to inject the environment variable. To my knowledge that is the single current use-case for the list, and we need to make sure that this proposed approach actually works with it.

We could use https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http for example, to retrieve the list and interpolate it into the environment variables for the API and/or ingestion server.

In any case, adding it to this implementation plan would allow us to document the implementation process for using this list rather than adding it to one of the other implementation plans that will be requested for the project (I suppose the "filtering" one?).

Author: I didn't realize we had a specific implementation in mind for getting the list into the repositories. That's awesome, I'll add this! And what a cool terraform data source.

Author: Added in 950bc718.

## This Will Not Be Perfect

By default, we accept that this list and its usage will not be perfect. At times
we may accidentally include terms which inadvertently cause harm to groups or
individuals. We may also accidentally expose users to undesired content in the
implementation of this list. Despite the potential for these transgressions this
work will always be based on the following assumptions:

1. Safety first. Given the choice, we would prefer to incorrectly restrict
   non-sensitive content rather than accidentally show sensitive content.
2. Avoid censorship. This list will not be used to block or deny access to
   content, rather to make the viewing of content opt-in. The Openverse project

>    will use other means to explicitly block or remove content. Removing

Reviewer: Is it worth clarifying that we would only block or remove content "in line with our content policy"? I assume that's following some WordPress Foundation content policy, but might be distinct, not sure.

Reviewer: I'm not sure either. Leaving it as ambiguous as possible (the way it is now) at least prevents us from saying something strictly incorrect. I'd like if we could ask someone familiar with WP Foundation's policies on this. Specifically, how would they like us to communicate the fact that we do reserve the right to remove content from Openverse for reasons other than DMCA and broad "illegality". (I totally agree that we need better understand how to communicate _that_ aspect as well.)

Author: Agreed that for now, we'll leave this intentionally ambiguous.

   sensitive content can, in many cases, be its own form of harm. A quote From
   the research paper "On the Dangers of Stochastic Parrots: Can Language Models
   Be Too Big? 🦜"[1] illustrates this well:

   “If we filter out the discourse of marginalized populations, we fail to
   provide training data that reclaims slurs and otherwise describes
   marginalized identities in a positive light”

3. Encourage feedback. We will request and monitor community feedback on the
   terms in this list (without inadvertently exposing anyone to the list) to
   make sure it meets the expectations of our users.

## Open questions for reviewers

- Within the Openverse maintainers; who has to view this list? Can anyone opt

>   out?

Reviewer: I think we should be able to opt out.

I'm also comfortable volunteering to review the list of words.

However, I will ask: what is the intention behind reviewing the list of words? Are we reviewing for format or to review the actual words themselves and make recommendations on their inclusion/exclusion?

Author: > Are we reviewing for format or to review the actual words themselves and make recommendations on their inclusion/exclusion?

I think both of these would be important but primarily reviewing the actual words themselves.

- With what frequency and urgency will community suggestions for the list be

>   reviewed, and by whom?

Reviewer: Will these be in GitHub issues? I wonder if we can round-robin the volunteers who opt-into reviewing (sort of like MSR, somehow) and have a process for pushing review off to someone else if a particular suggestion is too difficult for the assigned individual to handle on their own.

We should also identify what resources sponsored contributors have access to as far as counselling etc in dealing with potential repercussions from reviewing the list in the course of our work duties.

Author: > Will these be in GitHub issues? I wonder if we can round-robin the volunteers who opt-into reviewing (sort of like MSR, somehow) and have a process for pushing review off to someone else if a particular suggestion is too difficult for the assigned individual to handle on their own.

This is an excellent suggestion.

> We should also identify what resources sponsored contributors have access to

I'll look into this and update this section 👍

    - YouTube's
      [community guidelines](https://www.youtube.com/howyoutubeworks/policies/community-guidelines/)
      offers a decent list of categories for unacceptable types of content. For
      example, their
      [thumbnail policy](https://support.google.com/youtube/answer/9229980?hl=en&ref_topic=9282679#zippy=%2Cage-restricted-thumbnails-and-thumbnail-removal)
      includes some information on types of sexual and violent content that
      either can't be used or can be used behind a flag comparable to ours.
      - Additionally, please read
        [#this-will-not-be-perfect](#this-will-not-be-perfect).

## Implementation Plan

1. Ping a WordPress GitHub account maintainer to request that they create the
   `WordPress/openverse-sensitive-terms` repository.
2. Initialize the repository with a `README.md` and a blank
   `sensitive-terms.txt` file.
3. Create a PR which initializes the sensitive terms list by copying all terms
   in the "List of Dirty, Naughty, Obscene, and Otherwise Bad Words" repository.
   _I am not including the link here to prevent accidental viewing of the
   repository, but feel free to find it within the GitHub organization

>    "LDNOOBW"._

Reviewer 2: First time I see this list so I have to ask #sorryNotSorry :) What was used to select it and why? It's important so we can know how reliable it is. Can we get a bit of information on the State of the Art on these resources?

Author: Sure! I can add that.

# Implementation Plan: Managing A Sensitive Terms List

> **Content Warning**
>
> This document **does not** include sensitive content or direct links to
> sensitive content. It _does_ discuss different sub-categorizations of
> sensitive content at times.
>
> This document should not be considered sensitive but please view at your own
> discretion, regardless.

## Reviewers

<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->

- [x] @sarayourfriend

> - [ ] @dhruvkb

Reviewer 3: ```suggestion
- [x] @dhruvkb
```

3. Encourage feedback. We will request and monitor community feedback on the
   terms in this list (without inadvertently exposing anyone to the list) to
   make sure it meets the expectations of our users.

## Guidelines for using the list in Openverse projects

1. Never use terms from the list in Openverse repositories (issues, discussions,
   or code itself), chats, or communications of any time, outside of the
   sensitive terms repository.
2. Do not link directly to the list file itself; always reference the
   repository. The README will provide important context towards consenting to
   viewing the list.
3. Always use the most up-to-date version of the sensitive terms list when
   filtering and marking media as sensitive in Openverse projects.

## How the initial sensitive terms list was chosen

The initial set of terms, created by combining two lists:

- LDNOOBW/List-of-Dirty-Naughty-Obscene-and-Otherwise-Bad-Words
  - Very popular, used by many FOSS projects
  - Created explicitly for stock photography moderation
- surge-ai/profanity
  - Comprehensive
  - Contains more racist terms than LDNOOBW
  - Is categorized, which we're not using yet but could in the future

> Generally, its beneficial to us that our list be as comprehansive as possible.

Reviewer 4: ```suggestion
Generally, its beneficial to us that our list be as comprehensive as possible.
```

  - Contains more racist terms than LDNOOBW
  - Is categorized, which we're not using yet but could in the future

> Generally, its beneficial to us that our list be as comprehensive as possible.

Author: ```suggestion
Generally, it is beneficial to us that our list be as comprehensive as possible.
```
